pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,654 @@
+		blobs (instead of uploading). The source bucket also supports the suffix "/[store]", which 
+		will transform blob identifiers into the form the Docker registry uses on disk, allowing
+		you to mirror directly from an existing S3-backed Docker registry.
+
+		Images in manifest list format will be copied as-is unless you use --filter-by-os to restrict
+		the allowed images to copy in a manifest list. This flag has no effect on regular images.
+
+		Experimental: This command is under active development and may change without notice.`)
+
+	mirrorExample = templates.Examples(`
+# Copy image to another tag
+%[1]s myregistry.com/myimage:latest myregistry.com/myimage:stable
+
+# Copy image to another registry
+%[1]s myregistry.com/myimage:latest docker.io/myrepository/myimage:stable
+
+# Copy image to S3 (pull from <bucket>.s3.amazonaws.com/image:latest)
+%[1]s myregistry.com/myimage:latest s3://s3.amazonaws.com/<region>/<bucket>/image:latest
+
+# Copy image to S3 without setting a tag (pull via @<digest>)
+%[1]s myregistry.com/myimage:latest s3://s3.amazonaws.com/<region>/<bucket>/image
+
+# Copy image to multiple locations
+%[1]s myregistry.com/myimage:latest docker.io/myrepository/myimage:{stable,dev}
+
+# Copy multiple images
+%[1]s myregistry.com/myimage:latest=myregistry.com/other:test \ 
+    myregistry.com/myimage:new=myregistry.com/other:target
+`)
+)
+
+type DestinationType string
+
+var (
+	DestinationRegistry DestinationType = "docker"
+	DestinationS3       DestinationType = "s3"
+)
+
+type Mapping struct {
+	Source      imageapi.DockerImageReference
+	Destination imageapi.DockerImageReference
+	Type        DestinationType
+}
+
+type pushOptions struct {
+	Out, ErrOut io.Writer
+
+	Mappings []Mapping
+	OSFilter *regexp.Regexp
+
+	FilterByOS string
+
+	Insecure  bool
+	SkipMount bool
+	Force     bool
+
+	AttemptS3BucketCopy []string
+}
+
+// NewCommandMirrorImage copies images from one location to another.
+func NewCmdMirrorImage(name string, out, errOut io.Writer) *cobra.Command {
+	o := &pushOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "mirror SRC DST [DST ...]",
+		Short:   "Mirror images from one repository to another",
+		Long:    mirrorDesc,
+		Example: fmt.Sprintf(mirrorExample, name+" mirror"),
+		Run: func(c *cobra.Command, args []string) {
+			o.Out = out
+			o.ErrOut = errOut
+			kcmdutil.CheckErr(o.Complete(args))
+			kcmdutil.CheckErr(o.Run())
+		},
+	}
+
+	flag := cmd.Flags()
+	flag.BoolVar(&o.Insecure, "insecure", o.Insecure, "If true, connections may be made over HTTP")
+	flag.BoolVar(&o.SkipMount, "skip-mount", o.SkipMount, "If true, always push layers instead of cross-mounting them")
+	flag.StringVar(&o.FilterByOS, "filter-by-os", o.FilterByOS, "A regular expression to control which images are mirrored. Images will be passed as '<platform>/<architecture>'.")
+	flag.BoolVar(&o.Force, "force", o.Force, "If true, attempt to write all contents.")
```

Comment:
i'm confused on the overlap between Force and SkipMount.

Reply:
Skipmount may not work in all cases for openshift registries or older registries.  Since it's an optimization, if a user hits an issue I want them to be able to bypass its use.

Force says upload even if the image exists, i.e. don't trust the referential integrity of the registry.  If you had a broken prune, a mirror from a working registry would restore the blob.